File: python/src/fds/protobuf/stach/extensions/v1/StachUtilities.py
import math

from fds.protobuf.stach.NullValues import NullValues
from fds.protobuf.stach.table.DataType_pb2 import DataType


class StachUtilities:
    """ This class provides helper method for returning the data from the SeriesData object by handling the null values. """

    @staticmethod
    def get_value_helper(series_data, datatype, index, null_format):
        """
        The purpose of this function is to return the value from the provided SeriesData object.

        :param series_data: The data of a series. A series is how a column of data is represented.
        :param datatype: The type of data in a series.
        :param index: index can be referred as a key which is used to access the value inside list.
        :param null_format: null_format is used to render a null value with a special string, like -- or @NA.
        :return: Return data object from the SeriesData.
        """
        if DataType.Name(datatype) == "STRING":
            return StachUtilities.null_value_handler(datatype, series_data.string_array.values[index], null_format)
        elif DataType.Name(datatype) == "DOUBLE":
            return StachUtilities.null_value_handler(datatype, series_data.double_array.values[index], null_format)
        elif DataType.Name(datatype) == "FLOAT":
            return StachUtilities.null_value_handler(datatype, series_data.float_array.values[index], null_format)
        elif DataType.Name(datatype) == "INT32":
            return StachUtilities.null_value_handler(datatype, series_data.int32_array.values[index], null_format)
        elif DataType.Name(datatype) == "INT64":
            return StachUtilities.null_value_handler(datatype, series_data.int64_array.values[index], null_format)
        elif DataType.Name(datatype) == "BOOL":
            return StachUtilities.null_value_handler(datatype, series_data.bool_array.values[index], null_format)
        elif DataType.Name(datatype) == "DURATION":
            return StachUtilities.null_value_handler(datatype, series_data.duration_array.values[index], null_format)
        elif DataType.Name(datatype) == "TIMESTAMP":
            return StachUtilities.null_value_handler(datatype, series_data.timestamp_array.values[index], null_format)
        else:
            ValueError("The datatype is not implemented")

    @staticmethod
    def null_value_handler(datatype, value, null_format):
        """
        The purpose of this function is to handle the null values.

        :param datatype: The type of data in a series.
        :param value: Specifies a particular data of a series referenced by an index.
        :param null_format: null_format is used to render a null value with a special string, like -- or @NA.
        :return: Returns either null value with a special string, like -- or @NA or the value at the particular index.
        """
        if DataType.Name(datatype) == "STRING":
            if NullValues.STRING == value:
                return null_format
            return value
        elif DataType.Name(datatype) == "DOUBLE":
            if math.isnan(value):
                return null_format
            return value
        elif DataType.Name(datatype) == "FLOAT":
            if math.isnan(value):
                return null_format
            return value
        elif DataType.Name(datatype) == "INT32":
            if NullValues.INT32 == value:
                return null_format
            return value
        elif DataType.Name(datatype) == "INT64":
            if NullValues.INT64 == value:
                return null_format
            return value
        elif DataType.Name(datatype) == "DURATION":
            if NullValues.DURATION.equals(value):
                return null_format
            return value
        elif DataType.Name(datatype) == "TIMESTAMP":
            if NullValues.TIMESTAMP.equals(value):
                return null_format
            return value
        else:
            return value

File: dotnet/StachExtensions/FactSet.Protobuf.Stach.Extensions/V1/StachUtilities.cs
using System;
using FactSet.Protobuf.Stach.Table;

namespace FactSet.Protobuf.Stach.Extensions.V1
{
    public static class StachUtilities
    {
        /// <summary>
        /// The purpose of this function is to return the value from the provided SeriesData object.
        /// </summary>
        /// <exception cref="NotImplementedException">Thrown when datatype is not implemented</exception>
        /// <param name="seriesData"></param>
        /// <param name="dataType"></param>
        /// <param name="index"></param>
        /// <returns>Return data object from the SeriesData.</returns>
        public static object GetValueHelper(this SeriesData seriesData, DataType dataType, int index)
        {
            switch (dataType)
            {
                case DataType.Bool:
                    {
                        return seriesData.BoolArray?.Values?[index];
                    }
                case DataType.Double:
                    {
                        return seriesData.DoubleArray?.Values?[index];
                    }
                case DataType.Duration:
                    {
                        var v = seriesData.DurationArray?.Values?[index];
                        return v?.ToTimeSpan();
                    }
                case DataType.Float:
                    {
                        return seriesData.FloatArray?.Values?[index];
                    }
                case DataType.Int32:
                    {
                        return seriesData.Int32Array?.Values?[index];
                    }
                case DataType.Int64:
                    {
                        return seriesData.Int64Array?.Values?[index];
                    }
                case DataType.String:
                    {
                        return seriesData.StringArray?.Values?[index];
                    }
                case DataType.Timestamp:
                    {
                        var v = seriesData.TimestampArray?.Values?[index];
                        return v?.ToDateTime();
                    }
                default:
                    throw new NotImplementedException($"{dataType} is not implemented");
            }
        }
    }
}